x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTaskRandomTests.java

```diff
@@ -0,0 +1,282 @@
+                    .filter(Objects::nonNull)
+                    .flatMap(response -> Arrays.stream(response.getOperations()))
+                    .sorted(Comparator.comparingLong(Translog.Operation::seqNo))
+                    .collect(Collectors.toList());
+                assertThat(receivedOperations.size(), equalTo(expectedOperations.size()));
+                receivedOperations.sort(Comparator.comparingLong(Translog.Operation::seqNo));
+                for (int i = 0; i < receivedOperations.size(); i++) {
+                    Translog.Operation actual = receivedOperations.get(i);
+                    Translog.Operation expected = expectedOperations.get(i);
+                    assertThat(actual, equalTo(expected));
+                }
+            }
+        };
+    }
+
+    private static TestRun createTestRun(long startSeqNo, long startIndexMetadataVersion, int maxOperationCount) {
+        long prevGlobalCheckpoint = startSeqNo;
+        long indexMetaDataVersion = startIndexMetadataVersion;
+        int numResponses = randomIntBetween(16, 256);
+        Map<Long, List<TestResponse>> responses = new HashMap<>(numResponses);
+        for (int i = 0; i < numResponses; i++) {
+            long nextGlobalCheckPoint = prevGlobalCheckpoint + maxOperationCount;
+            if (sometimes()) {
+                indexMetaDataVersion++;
+            }
+
+            if (sometimes()) {
+                List<Translog.Operation> ops = new ArrayList<>();
+                for (long seqNo = prevGlobalCheckpoint; seqNo <= nextGlobalCheckPoint; seqNo++) {
+                    String id = UUIDs.randomBase64UUID();
+                    byte[] source = "{}".getBytes(StandardCharsets.UTF_8);
+                    ops.add(new Translog.Index("doc", id, seqNo, 0, source));
+                }
+
+                List<TestResponse> item = new ArrayList<>();
+                // Sometimes add a random retryable error
+                if (sometimes()) {
+                    Exception error = new UnavailableShardsException(new ShardId("test", "test", 0), "");
+                    item.add(new TestResponse(error, indexMetaDataVersion, null));
+                }
+                item.add(new TestResponse(null, indexMetaDataVersion,
+                    new ShardChangesAction.Response(indexMetaDataVersion, nextGlobalCheckPoint, ops.toArray(EMPTY))));
+                responses.put(prevGlobalCheckpoint, item);
+            } else {
+                // Simulates a leader shard copy not having all the operations the shard follow task thinks it has by
+                // splitting up a response into multiple responses AND simulates maxBatchSizeInBytes limit being reached:
+                long toSeqNo;
+                for (long fromSeqNo = prevGlobalCheckpoint; fromSeqNo <= nextGlobalCheckPoint; fromSeqNo = toSeqNo + 1) {
+                    toSeqNo = randomLongBetween(fromSeqNo, nextGlobalCheckPoint);
+                    List<Translog.Operation> ops = new ArrayList<>();
```

Comment:
Can we move this `ops` just before we generate the actual response (after the error response). It may be easier to follow.